File: frontend/admin/src/api/node.ts
import request from '@/utils/request'

// Node Interface
export interface Node {
    id: number
    name: string
    api_token: string
    endpoint: string
    location?: string
    country_code?: string
    status: string // 'online' | 'offline'
    cpu_usage: number
    memory_usage: number
    disk_usage: number
    bandwidth_usage: number
    instance_count?: number
    last_seen_at?: string
    created_at: string
}

// Create Node Request
export interface CreateNodeRequest {
    name: string
    endpoint: string
    location?: string
    country_code?: string
}

// Update Node Request
export interface UpdateNodeRequest {
    name?: string
    endpoint?: string
    location?: string
    country_code?: string
}

// Get Node List
export function getNodeList() {
    return request<Node[]>({
        url: '/admin/nodes',
        method: 'get'
    })
}

// Create Node
export function createNode(data: CreateNodeRequest) {
    return request<Node>({
        url: '/admin/nodes',
        method: 'post',
        data
    })
}

// Get Node Detail
export function getNode(id: number) {
    return request<Node>({
        url: `/admin/nodes/${id}`,
        method: 'get'
    })
}

// Update Node
export function updateNode(id: number, data: UpdateNodeRequest) {
    return request<Node>({
        url: `/admin/nodes/${id}`,
        method: 'put',
        data
    })
}

// Delete Node
export function deleteNode(id: number) {
    return request({
        url: `/admin/nodes/${id}`,
        method: 'delete'
    })
}

// Regenerate Token
export function regenerateToken(id: number) {
    return request<{ token: string }>({
        url: `/admin/nodes/${id}/token`,
        method: 'post'
    })
}

// Download Install Script
export function downloadInstallScript(id: number, nodeName: string) {
    return request({
        url: `/admin/nodes/${id}/install-script`,
        method: 'get',
        responseType: 'blob'
    }).then((blob: Blob) => {
        // Create download link
        const url = window.URL.createObjectURL(blob)
        const link = document.createElement('a')
        link.href = url
        link.download = `install-agent-${nodeName}.sh`
        document.body.appendChild(link)
        link.click()
        document.body.removeChild(link)
        window.URL.revokeObjectURL(url)
    })
}

File: backend/master/internal/model/node.go
package model

import "time"

// Node 表示 Snell 节点。
type Node struct {
	ID             uint       `gorm:"primaryKey" json:"id"`
	Name           string     `gorm:"uniqueIndex;size:64;not null" json:"name"`
	APIToken       string     `gorm:"uniqueIndex;size:128;not null" json:"api_token"`
	Endpoint       string     `gorm:"size:255;not null" json:"endpoint"`
	Location       string     `gorm:"size:100" json:"location"`
	CountryCode    string     `gorm:"size:8" json:"country_code"`
	Status         string     `gorm:"size:32;default:'offline'" json:"status"`
	CPUUsage       float64    `gorm:"default:0" json:"cpu_usage"`
	MemoryUsage    float64    `gorm:"default:0" json:"memory_usage"`
	DiskUsage      float64    `gorm:"default:0" json:"disk_usage"`
	BandwidthUsage float64    `gorm:"default:0" json:"bandwidth_usage"`
	LastSeenAt     *time.Time `json:"last_seen_at"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`

	Users     []User          `gorm:"many2many:user_nodes" json:"users,omitempty"`
	Instances []SnellInstance `gorm:"foreignKey:NodeID" json:"instances,omitempty"`
}
